opm/simulators/linalg/gpuistl/set_device.cpp

```diff
@@ -43,7 +44,28 @@ setDevice(int mpiRank, [[maybe_unused]] int numberOfMpiRanks)
     const auto deviceId = mpiRank % deviceCount;
     OPM_GPU_SAFE_CALL(cudaDeviceReset());
     OPM_GPU_SAFE_CALL(cudaSetDevice(deviceId));
-    OpmLog::info("Set CUDA device to " + std::to_string(deviceId) + " (out of " + std::to_string(deviceCount)
-                 + " devices).");
 }
+
+void
+printDevice (int mpiRank, int numberOfMpiRanks)
+{
+    int deviceCount = -1;
+    OPM_GPU_SAFE_CALL(cudaGetDeviceCount(&deviceCount));
+
+    const auto deviceId = mpiRank % deviceCount;
```

Comment:
This looks like there is an association of MPI ranks to GPU devices.
Out of curiosity: How are we enforcing this associating?

Basically, we are assuming that e.g. ranks 0,4,8,12,... are on the same node if we have for gpus in the calculation. There is of course some kind of control where to start MPI ranks. But it is actually really hard to do this. So the default would be that we need to detect which ranks are near the GPU (read on the same node) to reduce latency. Probably there is some machanism for that...

Reply:
Having the rank modulo the local device count would mean mapping for instance ranks {0, 1, 2, 3} to the same node if we have four gpus that we want to use, not {0, 4, 8, 12}? In the latter case the result of the modulo operation would always be 0 and these four processes would just use device0.

Reply:
My bad. I should give back my degrees.

My point was  that I am doubting that one can assume any locations for the MPI ranks at all (unless enforcing it from the outside). Hence a rank associated with a GPU might live on a distant node, if if the rank is close. Depending on the interconnect that might make a difference. But I guess you are aware about this.

Seems like you can actually determine local ranks via creating communicators with MPI_COMM_TYPE_SHARED. The you can  get an "on-node" communicator, to do the association in a multi-node-safe manner. See https://medium.com/@jeffrey_91423/binding-to-the-right-gpu-in-mpi-cuda-programs-263ac753d232

Reply:
For GPU the only thing that makes sense I think is to use N processes, where N is the number of GPUs (Nodes * gpusPerNode), and then I think filling up the processes in a consecutive way (node1 has {p0, p1, p2, p3}, node2 has {p4, p4...}... will work just fine for now. This is probably not that hard to do with slurm, which I hope we will use in the multigpu case in the foreseeable future? Otherwise I can create an issue requesting more functionality to select gpus so we keep this in mind and handle it later.

Reply:
Just to add to this: The current way we are assigning the GPUs is very rudimentary, and might not yield the best performance (especially for two socket/CPU NUMA server nodes with multiple GPUs), nor is it guaranteed that the MPI ranks will map that nicely to different nodes/GPUs unless you control the submission. This is also why I wrote:

```C++
   // Now do a round robin kind of assignment
   // TODO: We need to be more sophistacted here. We have no guarantee this will pick the correct device.
```
(typo is highly intentional!)

The simple reason why this hasn't been developed or investigated further because we are not really at a point where running over multiple nodes with multiple GPUs is relevant. However, we are applying for access to LUMI-G, where this will be looked into. The current solution "works well enough" for the current testing we are doing.